### ISSUE TITLE
Android pre API 24 language support

### ISSUE THREAD

**Issue by snodnipper** [wdtinc/mapbox-vector-tile-java#13]
I have spent the day running code quality tools to get this going on old versions of Android.  I can feedback as a separate issue - I needed Java 1.7 API compatibility, for instance.

Anyhow, these lines fall through and look suspicious.

https://github.com/wdtinc/mapbox-vector-tile-java/blob/fc61116063e6b9fa072a26cd73146920edd0e473/src/main/java/com/wdtinc/mapbox_vector_tile/adapt/jts/JtsAdapter.java#L445

https://github.com/wdtinc/mapbox-vector-tile-java/blob/fc61116063e6b9fa072a26cd73146920edd0e473/src/main/java/com/wdtinc/mapbox_vector_tile/adapt/jts/JtsAdapter.java#L517

https://github.com/wdtinc/mapbox-vector-tile-java/blob/master/src/main/java/com/wdtinc/mapbox_vector_tile/adapt/jts/MvtReader.java#L535

https://github.com/wdtinc/mapbox-vector-tile-java/blob/master/src/main/java/com/wdtinc/mapbox_vector_tile/adapt/jts/MvtReader.java#L592

**Comment by ShibaBandit** [comment]
I haven't done much android in a while, but I heard they support java 8 via desugar and deprecated Jack? Would this be valid on old android versions? How far back are you needing to go?
https://developer.android.com/studio/write/java8-support.html

https://developer.android.com/reference/java/util/Collections.html#emptyList()
https://developer.android.com/reference/java/util/List.html#toArray()

**Comment by snodnipper** [comment]
```java
-    final Integer mapIndex = keys.putIfAbsent(key, nextIndex);
+    // Android API 24 call
+    // final Integer mapIndex = keys.putIfAbsent(key, nextIndex);
+    final Integer mapIndex;
+    if (!vals.containsKey(key)) {
+      mapIndex = vals.put(key, nextIndex);
+    }
+    else {
+      mapIndex = vals.get(key);
+    }
```

Minor things like this.  It is not pretty _but_ is documented.  https://developer.android.com/reference/java/util/Map.html#putIfAbsent(K, V)

It certainly isn't a pleasurable experience...because usually Java version defines allowed syntax and methods that can be called.

Continuing...

**Comment by ShibaBandit** [comment]
Added travis ci to run 'mvn clean test' to master.

**Comment by ShibaBandit** [comment]
Created branch 'android_api_24_support'

**Comment by snodnipper** [comment]
Sweet - I'll commit the changes currently visible in our [vt-support lib](https://github.com/OrdnanceSurvey/vt-support/tree/5e2f678c51d4f6d99abd0d5ceab3f5b39a4c6ae0/vt-java-legacy-parser) and start a PR.

**Comment by ShibaBandit** [comment]
@snodnipper Since you've tested with API 21 mentioned in #14 I may go ahead and update the readme to reflect that.

**Comment by ShibaBandit** [comment]
Closed via #14 and #16

**Comment by snodnipper** [comment]
Thanks @ShibaBandit and folks for commenting etc. to help push this over the line.

If anyone wants to talk about functionality etc. I am happy to talk on the http://thespatialcommunity.org/ slack channel - e.g. very old Android support etc.

👍
